torch/csrc/jit/codegen/cuda/scheduler.cpp

```diff
@@ -0,0 +1,243 @@
+#include <torch/csrc/jit/codegen/cuda/scheduler.h>
+
+#include <torch/csrc/jit/codegen/cuda/arith.h>
+#include <torch/csrc/jit/codegen/cuda/ir_all_nodes.h>
+#include <torch/csrc/jit/codegen/cuda/ir_iostream.h>
+#include <torch/csrc/jit/codegen/cuda/parser.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+constexpr int kUnrollFactor = 4;
+
+namespace {
+
+std::vector<int> reductionAxes(TensorView* tv) {
+  size_t n_dims = tv->nDims();
+  std::vector<int> reduction_axes;
+  for (size_t i = 0; i < n_dims; i++) {
+    if (tv->axis(i)->isReduction()) {
+      reduction_axes.emplace_back(i);
+    }
+  }
+  return reduction_axes;
+}
+
+// coalesces all reduction to the right side and returns total number of
+// reduction axes
+size_t coalescReduction(TensorView* tv) {
+  auto reduction_axes = reductionAxes(tv);
+  size_t n_dims = tv->nDims();
+  std::unordered_map<int, int> coalesc_permute;
+  for (size_t i = 0; i < reduction_axes.size(); i++) {
+    size_t new_pos = i + n_dims - reduction_axes.size();
+    if (new_pos == reduction_axes[i]) {
+      break;
+    } else {
+      coalesc_permute[reduction_axes[i]] = new_pos;
+    }
+  }
+  if (!coalesc_permute.empty()) {
+    tv->reorder(coalesc_permute);
+  }
+  return reduction_axes.size();
+}
+
+} // namespace
+
+// This one is a total mess and it should go.
+bool scheduleFusion(Fusion* fusion, const at::ArrayRef<c10::IValue> inputs) {
```

Comment:
The implementation here is not relevant. @kevinstephano 's new scheduler should be swapped in here at some point.
This is going to be removed soon-ish, so I don't see much value in refactoring this to separate the module for PW/fcd_reduction/non_fcd_reduction.